src/storage/src/types/sources.rs

```diff
@@ -1523,25 +1528,119 @@ impl Generator {
+            generator: self.clone(),
+            invocation: 0,
+            pending: VecDeque::new(),
+            rng: SmallRng::seed_from_u64(seed.unwrap_or_default()),
+            now,
         }
     }
 }
 
-impl FromStr for Generator {
-    type Err = String;
+pub struct GeneratorIterator {
+    generator: Generator,
+    // Number of times `pending` has been refilled.
+    invocation: i64,
+    pending: VecDeque<Row>,
+    rng: SmallRng,
+    now: NowFn,
+}
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "counter" => Ok(Self::Counter),
-            _ => Err("unknown load generator".into()),
+impl Iterator for GeneratorIterator {
+    type Item = Row;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.pending.is_empty() {
+            self.invocation += 1;
+            let pending = match self.generator {
+                Generator::Counter => {
+                    VecDeque::from([Row::pack_slice(&[Datum::Int64(self.invocation)])])
+                }
+                Generator::Auction => {
+                    const AUCTIONS: &[&str] = &[
+                        "Signed Memorabilia",
+                        "City Bar Crawl",
+                        "Best Pizza in Town",
+                        "Gift Basket",
+                        "Custom Art",
+                    ];
+
+                    let now = to_datetime((self.now)());
+                    let mut auction = Row::with_capacity(2);
+                    let mut packer = auction.packer();
+                    packer.push(Datum::String("auctions"));
+                    packer.push_list(&[
+                        Datum::String(&self.invocation.to_string()), // auction id
+                        Datum::String(AUCTIONS.choose(&mut self.rng).unwrap()), // item
+                        Datum::String(&(now + chrono::Duration::seconds(10)).to_string()), // end time
+                    ]);
+                    let mut rows = VecDeque::from([auction]);
+                    const MAX_BIDS: i64 = 10;
+                    for i in 0..self.rng.gen_range(2..MAX_BIDS) {
+                        let mut bid = Row::with_capacity(2);
+                        let mut packer = bid.packer();
+                        packer.push(Datum::String("bids"));
+                        packer.push_list(&[
+                            Datum::String(&(self.invocation * MAX_BIDS + i).to_string()), // bid id
+                            Datum::String(&self.invocation.to_string()), // auction id
+                            Datum::String(&self.rng.gen_range(1..100).to_string()), // amount
+                            Datum::String(&(now + chrono::Duration::seconds(i)).to_string()), // bid time
+                        ]);
+                        rows.push_back(bid);
+                    }
+                    rows
+                }
+            };
+            self.pending = pending;
         }
+        self.pending.pop_front()
```

Comment:
Since it's likely that we'll have many of these load generators of increasing complexity, I'd love to plan for that now and have a separate module for each load generator source. This file is already too long!

Reply:
+1. What would be really great is a single trait, so I or devex could be more involved

Reply:
Unfortunately we've tried and failed to make that trait for several years. 🙈